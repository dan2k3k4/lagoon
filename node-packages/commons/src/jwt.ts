import { sign } from 'jsonwebtoken';
type Role = 'none' | 'admin' | 'drush';

interface Payload {
  role: Role;

  // Issuer - Information on who created this token
  iss: string;

  // Subject - Unique identifier for the user / identity
  sub?: string;

  // Audience - Url of the resource server the token was intended for (e.g. http://api.amazee.io)
  aud?: string;

  // IssuedAt timestamp (needs to be > 0)
  iat?: number;

  // Expiration as NumericDate
  // Example for an expiration timestamp in 1 hour: Math.floor(Date.now() / 1000) + (60 * 60)
  exp: number;
}

// Create a JWT without any user ID checks.
//
// This function is also synchronous, which is useful for creating service
// tokens, which generally get created during startup of services. Beware
// the performance implications of using this in async environments!
export const createJWTWithoutUserId = (
  payload: Payload,
  jwtSecret: string,
): string => sign(payload, jwtSecret);
