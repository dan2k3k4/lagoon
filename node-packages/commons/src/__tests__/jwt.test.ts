import jwt from 'jsonwebtoken';
import { createJWTWithoutUserId } from '../jwt';

describe('createJWT', () => {
  test('should create token properly', async () => {
    const { payload, jwtSecret } = {
      payload: {
        role: 'admin' as const,
        iat: 1503014400000,
        exp: 1503014400000,
        iss: 'jest'
      },
      jwtSecret: 'secret',
    };

    const token = await createJWTWithoutUserId(payload, jwtSecret);
    const decodedPayload = jwt.verify(token, jwtSecret);

    expect(token).toMatchSnapshot();
    expect(decodedPayload).toEqual(payload);
  });

  test('should create token with expiration date via expiresIn argument', async () => {
    const hour = 3600;
    const currentTime = Math.floor(Date.now() / 1000);

    const { payload, jwtSecret } = {
      payload: {
        role: 'admin' as const,
        iat: currentTime,
        exp: currentTime + hour,
        iss: 'jest',
      },
      jwtSecret: 'secret',
    };

    const token = await createJWTWithoutUserId(payload, jwtSecret);
    const decodedPayload = jwt.verify(token, jwtSecret);
    const exp = typeof decodedPayload === 'object' ? decodedPayload.exp : 'jwt verify error';

    expect(exp).toEqual(currentTime + hour);
  });

  test('should create expired token, which should be reckognized by verify()', async () => {
    const { payload, jwtSecret } = {
      payload: {
        userId: 0,
        permissions: {
          projects: [],
          customers: []
        },
        role: 'admin' as const,
        iat: 1503014400, // way back in the past
        exp: 1503018000,
        iss: 'jest',
      },
      jwtSecret: 'secret',
    };

    const token = await createJWTWithoutUserId(payload, jwtSecret);
    const fn = () => jwt.verify(token, jwtSecret);

    expect(fn).toThrow(new Error('jwt expired'));
  });
});
