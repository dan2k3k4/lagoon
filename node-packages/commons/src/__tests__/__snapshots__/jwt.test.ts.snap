// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJWT should create token properly 1`] = `"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJyb2xlIjoiYWRtaW4iLCJpYXQiOjE1MDMwMTQ0MDAwMDAsImV4cCI6MTUwMzAxNDQwMDAwMCwiaXNzIjoiamVzdCJ9.LBa6msnHJEDbv26ynE_Q07E8ZSJH7PuOIjqUhBSPhQI"`;
